model/jpa/src/main/java/org/keycloak/authorization/jpa/store/JPAPolicyStore.java

```diff
@@ -285,6 +287,80 @@ public void findByScopes(ResourceServer resourceServer, Resource resource, List<
                 .forEach(consumer::accept);
     }
 
+    @Override
+    public void findResourcePermissionPolicies(ResourceServer resourceServer, Resource resource, Collection<Scope> scopes, String resourceType, boolean resourceServerPolicies, Consumer<Policy> consumer) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<PolicyEntity> policyQuery = cb.createQuery(PolicyEntity.class);
+        Root<PolicyEntity> root = policyQuery.from(PolicyEntity.class);
+        policyQuery.select(root);
+
+        List<Predicate> predicates = new LinkedList<>();
+
+        Join<PolicyEntity, ResourceEntity> resourceJoin = root.join("resources", JoinType.LEFT);
+        Join<PolicyEntity, ScopeEntity> scopeJoin = root.join("scopes", JoinType.LEFT);
+        MapJoin<PolicyEntity, String, String> policyConfigJoin = root.joinMap("config", JoinType.LEFT);
+
+        // selection on resource id provided. Will also naturally overlap with scopes and types needed
+        if(resource != null) {
+            predicates.add(cb.equal(resourceJoin.get("id"), resource.getId()));
+        }
+
+        // selection on scope ids with no resource
+        if (CollectionUtil.isNotEmpty(scopes)) {
+            Predicate baseScopePredicate = scopeJoin.get("id").in(scopes.stream().map(Scope::getId).collect(Collectors.toList()));
+            baseScopePredicate = cb.and(
+                    baseScopePredicate,
+                    cb.isEmpty(root.get("resources")),
+                    // the named query excludes any policy that has a resourceType, not sure that this is expected behavior
```

Comment:
reason this is needed is because the type queries will grab results without regarding scopes. Overlap between 
(all resource matches) + (type & no resources) + (scope & no resources & no types) 
is all relevant policies. With the transaction setup this way we can also selectively query just the parts we're missing in the cache too